api/python/tests/integration/test_packages.py

```diff
@@ -1410,3 +1423,106 @@ def test_resolve_hash(self):
 
         with pytest.raises(QuiltException, match='Found multiple matches'):
             Package.resolve_hash(pkg_name, LOCAL_REGISTRY, hash_prefix)
+
+
+# The following tests were moved out of the PackageTest class to enable parametrization.
+# see (https://docs.pytest.org/en/latest/unittest.html#pytest-features-in-unittest-testcase-subclasses)
+@pytest.mark.parametrize(
+    'target_dir, update_policy, expected_one_byte, expected_two_byte, expected_three_byte, expected_keys',
+    [
+        ('/', None, b'one', b'two', b'three', {'one.txt', 'two.txt', 'three.txt', 'sub'}),
+        ('/', 'incoming', b'one', b'two', b'three', {'one.txt', 'two.txt', 'three.txt', 'sub'}),
+        ('/', 'existing', b'1', b'two', b'three', {'one.txt', 'two.txt', 'three.txt', 'sub'}),
+        ('', 'incoming', b'one', b'two', b'three', {'one.txt', 'two.txt', 'three.txt', 'sub'}),
+        ('', 'existing', b'1', b'two', b'three', {'one.txt', 'two.txt', 'three.txt', 'sub'}),
+        ('sub/', 'incoming', b'one', b'two', b'three', {'one.txt', 'sub'}),
+        ('sub/', 'existing', b'one', b'2', b'3', {'one.txt', 'sub'}),
+        ('new-sub/', 'incoming', b'one', b'two', b'three', {'one.txt', 'sub', 'new-sub'}),
+        ('new-sub/', 'existing', b'one', b'two', b'three', {'one.txt', 'sub', 'new-sub'}),
+        pytest.param('/', 'bad_policy', b'1', b'2', b'3', set(), marks=pytest.mark.xfail(raises=ValueError)),
+    ]
+)
+def test_set_dir_update_policy(
+    target_dir: str,
+    update_policy: str,
+    expected_one_byte: bytes,
+    expected_two_byte: bytes,
+    expected_three_byte: bytes,
+    expected_keys: set
+):
+    """Verify building a package with update policy. """
+    nested_dir = DATA_DIR / 'nested'
+    pkg = Package()
+    pkg.set_dir("/", nested_dir, meta={'name': 'test_meta'})
+    assert set(pkg.keys()) == {'one.txt', 'sub'}
+    assert set(pkg['sub'].keys()) == {'two.txt', 'three.txt'}
+    assert pkg.meta == {'name': 'test_meta'}
+
+    nested_dir_2 = DATA_DIR / 'nested2'
+    if update_policy:
+        pkg.set_dir(target_dir, nested_dir_2, update_policy=update_policy)
+    else:
+        pkg.set_dir(target_dir, nested_dir_2)
+    assert set(pkg.keys()) == expected_keys
+
+    target_dir = target_dir.strip("/")
+    if target_dir:
```

Comment:
ideally we wouldn't need this branch because the parametrize expectations are all self-contained, but this is fine for now.